### PR TITLE
Structural Error Cases for Lambda Default Parameters

### LINKED ISSUE #62960: Lambda parameters are not checked for a number of error conditions
```cs
Delegate d = (ref TypedReference typedReference) => 1;
```

This should error with `error CS1601: Cannot make reference to variable of type 'TypedReference'`. It does not, and compiles successfully.

### PR DESCRIPTION
**Pull request by adamperlin**
This is an incremental PR that adds the following error cases in the binder for lambda default parameters:

- Optional parameter before required parameter, i.e., `(int x = 3, int y) => x + y`
- Implicitly typed optional parameter, ex: `(x = 10) => x`
- Type mismatch in default parameter value, ex:  `(string s = 3) => s`
- Non-constant default parameter value, ex:  `(C c = new C()) => c`

The following is currently **Not** handled in the binder:
- Inferring the proper anonymous delegate type for lambdas with default parameters
- Any error cases related to assignment / delegate conversion for example, default value mismatch when a lambda is assigned to another lambda

Also, the refactoring around lambda parameter error checking resolves #62960.

Relates to test plan https://github.com/dotnet/roslyn/issues/62485

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs:373]: We'll need to talk about this a bit. I'm concerned with spreading these checks out more, as we already have a decent amount of duplication of the checks in `ParameterHelpers.ReportParameterErrors` (to our detriment, as we miss at least one error scenario from that method here).

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:544]: Stray whitespace

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:606]: Stray whitespace

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:12]: Unnecessary?

File: src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs:12]: Unnecessary?

File: src/Compilers/CSharp/Portable/CSharpResources.resx
Comment (333fred) [inline src/Compilers/CSharp/Portable/CSharpResources.resx:7188]: Extra leading and trailing spaces. Also, while we aren't perfectly consistent on this (see the above message...) we do generally try to end errors with a `.`. Finally, `Default` isn't really a good descriptor. Consider something more like `Implicitly typed lambda parameter '{0}' cannot have a default value.`

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:635]: We should avoid realizing this if we don't need to.

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:403]: We shouldn't use a catch-all. We should list the explicit cases we expect to see, and throw `ExceptionUtilities.UnexpectedValue` for cases we didn't expect to see.

File: src/Compilers/CSharp/Portable/Symbols/Source/LambdaParameterSymbol.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/LambdaParameterSymbol.cs:26]: `HasDefaultArgumentSyntax` below should probably be updated. We should look through the users of that API and make sure they're doing the right thing.

**Review by 333fred — COMMENTED** [review]
Done with review pass for now. I haven't gone through the tests yet.

I'm concerned with leaving the default parameter binding to `SourceComplexParameterSymbol`. It doesn't have access to the binder from the `UnboundLambda`, and it seems likely to get something wrong. We should make sure we have tests both in and out of checked containing contexts, nested inside containing lambdas and referencing their parameter names in default parameters, and other things in the area. As a concrete example, I expect that this will currently error, but should be legal:

```cs
void M()
{
    unchecked
    {
        _ = (int i = int.MaxValue + 1) {}
    }
}
```

**Comment by 333fred** [comment]
Also, test failures look legitimate.

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:673]: This isn't the correct location. We'll need to start by filling the array up to the current parameter with nulls.

Comment (adamperlin) [reply src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:673]: Ah good catch. And front-filling up to the current parameter should be sufficient since at this point we should never have a case where the array needs to look like `{null, ..., <value>, ..., null}` as that would imply that default params came before required ones

Comment (adamperlin) [reply src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:673]: I realize I don't have any test cases for this code yet because I hadn't gotten to the point in code yet where we would actually use this array

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:551]: >ParamSyntax [](http://example.com/codeflow?start=41&length=11)

Consider spelling out "Parameter": `ParameterSyntax(int i)`

File: src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs:262]: >Implicitly Typed Default Parameters [](http://example.com/codeflow?start=19&length=35)

Consider using lowercase within the sentence, here and in the matching comment below: "Implicitly typed default parameters ..."

File: src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs:278]: >parameter.Default != null [](http://example.com/codeflow?start=54&length=25)

Minor: This could be inlined in the one use below, to avoid getting the default unnecessarily.

File: src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs:289]: >// Error here. [](http://example.com/codeflow?start=28&length=14)

The comment is not needed.

File: src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs:317]: >UNDONE [](http://example.com/codeflow?start=23&length=6)

Consider using `PROTOTYPE` in the feature branch instead, since we automatically catch `PROTOTYPE` comments before merging into main.

File: src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs:310]: >paramSyntax should not be null here [](http://example.com/codeflow?start=23&length=35)

Let's add `Debug.Assert(paramSyntax is { });`, to avoid needing `paramSyntax!` below.

File: src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs:312]: >paramSyntax!.Default [](http://example.com/codeflow?start=24&length=20)

If we're using `paramSyntax.Default` directly, do we need `UnboundLambda.DefaultValue(int)`?

File: src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs:151]: >p.Default [](http://example.com/codeflow?start=44&length=9)

Where are we binding these values?

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:399]: >var paramSyntaxList = syntax switch [](http://example.com/codeflow?start=12&length=35)

I'm not sure if we should be binding default values here (within UnboundLambda) or in the caller. (We should probably decide that first.) But if it's here and we need the syntax for individual parameters, should the caller provide that directly?

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:401]: >Lam [](http://example.com/codeflow?start=69&length=3)

Consider using "Lambda" here and in other locals.

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:474]: >ParameterDefaultValue [](http://example.com/codeflow?start=40&length=21)

Is this used?

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:476]: >ParameterHasDefault [](http://example.com/codeflow?start=20&length=19)

Is this used?

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:674]: >paremeters [](http://example.com/codeflow?start=75&length=10)

Typo.

File: src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs
Comment (cston) [reply src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs:151]: It looks like binding occurs in `SourceComplexParameterSymbolBase`.

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:1494]: >_defaultValues.IsDefault || (0 <= index && index < _defaultValues.Length) [](http://example.com/codeflow?start=25&length=73)

Consider strengthening the assert, and matching the other methods above, by comparing `index` with `_parameterTypesWithAnnotations.Length` unconditionally.

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:610]: >owner != null [](http://example.com/codeflow?start=33&length=13)

We typically use reference equality when checking a `Symbol` for `null`. Consider using `owner is { }` or `(object)owner != null`.

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:590]: >TypeSymbol paramType [](http://example.com/codeflow?start=12&length=20)

Consider removing this parameter and replacing with `paramTypeWithAnnotations.Type`.

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:593]: >paramContainingSymbol [](http://example.com/codeflow?start=20&length=21)

Is this different from `owner`? If not, consider removing.

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:589]: >param [](http://example.com/codeflow?start=32&length=5)

Consider spelling out "parameter" for each of these parameters, or better yet, dropping the "param" prefix if the meaning is unambiguous.

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:648]: >ReportParameterErrors [](http://example.com/codeflow?start=28&length=21)

It looks like there is only one caller. Consider inlining this method there.

File: src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs:7166]: >LambdaWithDefaultTypeMismatchConstantExpression [](http://example.com/codeflow?start=20&length=47)

The type mismatch cases above seem sufficient. These two can probably be removed.

File: src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs:317]: This is copied from the original implementation, it's not newly introduced in this branch.

Comment (adamperlin) [reply src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs:317]: I'm happy to change this though if a different flag would be better

Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs:317]: No, it's not a new thing, so let's leave it as is. We have no plans to address it.

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs
Comment (adamperlin) [reply src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:593]: The lambda code doesn't make use of either of these arguments. However, when `MakeParameters` calls this function, it passes through the `owner` parameter it got passed, plus the containing symbol attached to a particular parameter (`ParameterSymbol.ContainingSymbol`). I don't have the knowledge to say if those would always be the same

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs
Comment (adamperlin) [reply src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:399]: The unbound lambda depends on the syntax node already, so it seemed better to just make use of the respective properties from the underlying syntax node

Comment (cston) [reply src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:399]: It looks like there's only one caller of `UnboundLambda.Create()` and that caller already switches on the anonymous function syntax kind to get other values. I think for consistency, we should handle the parameter syntax there.

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs
Comment (cston) [reply src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:593]: Ok, I think it is the same as `owner` but it seems fine to pass in as a separate parameter. Perhaps, call it `containingSymbol` though. :)

File: src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs:7327]: Let's leave a PROTOTYPE comment that these warnings should go away.

File: src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs:7182]: Let's add a potential cycle test: `var lam = (Delegate d = lam) => { }`.

File: src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs:7393]: There's no params in this test (and the next one)?

Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs:7393]: Oh, I see, you using it as a shorthand. Consider spelling out Parameters, as `params` is a C# keyword we'll be testing at some point.

File: src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs:7449]: Please leave a PROTOTYPE comment to verify this with the DataflowAnalysis APIs.

**Review by 333fred — COMMENTED** [review]
Done review pass (commit 10). Implementation is looking good, but there are a few test adjustments to make.

**Review by 333fred — APPROVED** [review]
LGTM (commit 28). @cston for the second review.

File: src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs:266]: Consider extracting to a local function, for reuse here and in the loop below.

File: src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs:267]: Consider extracting to a local function, for reuse here and in the loop below.

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:664]: >param.ExplicitDefaultConstantValue [](http://example.com/codeflow?start=53&length=34)

`constVal`

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:668]: >parameterDefaultValues [](http://example.com/codeflow?start=16&length=22)

Is this array used?

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs
Comment (cston) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs:3238]: >CS1065ERR_DefaultValueNotAllowed [](http://example.com/codeflow?start=20&length=32)

Unless there's a good reason, we should keep existing tests, even if the behavior has changed, so that we're confirming the current expected behavior.

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs
Comment (adamperlin) [reply src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:668]: > > parameterDefaultValues
> 
> Is this array used?

Not currently, but it will be in my next PR. I decided on this as a cut-off point for this PR

**Review by cston — APPROVED** [review]

**Pull request merged**
